### ISSUE TITLE
Difficulties with Hdf5_caml

### ISSUE THREAD

**Issue by mars0i** [vbrankov/hdf5-ocaml#10]
This continues a discussion that started in #3.

OK, I don't understand why I can't `#require "hdf5_caml"` in utop if `caml_copy_string` is supposed to be available at the top level; it's not.  I've posted a question about on discuss.ocaml.org.

I copied a couple of the example files to a project and listed them in a jbuild file.  I can compile `examples/h5_intro/create_example.ml` into a native executable that runs just fine, although there's a warning about the missing `serial` directory that I mentioned in #3.  This example file uses `Hdf5_raw`. 

I also copied `examples/caml/test.ml` to my project.  This uses `Hdf5_caml`.  When I add that to the jbuild file, it causes an error that I find very puzzling.

jbuild:
```
(jbuild_version 1)

(executables
 ((names (create_example test))
	(public_names (create_example test))
  (package yo)
  (libraries (bigarray hdf5_raw hdf5_caml))
 ))
```
(The version that worked with `create_example.ml` is the same, but with both instances of "test" deleted and with "hdf5_caml" deleted.)

The error:
```
$ make
jbuilder build @install
      ocamlc src/yo/test.{cmi,cmo,cmt}
      ocamlc src/yo/create_example.{cmi,cmo,cmt}
    ocamlopt src/yo/test.{cmx,o}
    ocamlopt src/yo/test.exe (exit 2)
(cd _build/default && /Users/marshall/.opam/4.05.0+flambda/bin/ocamlopt.opt -w -40 -g -o src/yo/test.exe -I /Users/marshall/.opam/4.05.0+flambda/lib/hdf5_caml -I /Users/marshall/.opam/4.05.0+flambda/lib/hdf5_raw -I /Users/marshall/.opam/4.05.0+flambda/lib/ocaml /Users/marshall/.opam/4.05.0+flambda/lib/ocaml/unix.cmxa /Users/marshall/.opam/4.05.0+flambda/lib/hdf5_raw/hdf5_raw.cmxa /Users/marshall/.opam/4.05.0+flambda/lib/ocaml/bigarray.cmxa /Users/marshall/.opam/4.05.0+flambda/lib/hdf5_caml/hdf5_caml.cmxa src/yo/test.cmx)
File "_none_", line 1:
Error: No implementations provided for the following modules:
         Bigarray referenced from /Users/marshall/.opam/4.05.0+flambda/lib/hdf5_raw/hdf5_raw.cmxa(Hdf5_raw)
make: *** [all] Error 1
```
This seems strange because test.ml makes no reference to `Bigarray`, and I am including it in the `libraries` entry of jbuild.

Sorry to raise all of these issues at once.  I'll keep working on them, but was wondering whether you had any insight.  Thanks.

**Comment by mars0i** [comment]
The `(cd _build/default && ... ocamlopt.opt ...)` line above lists `... .opam/4.05.0+flambda/lib/ocaml/bigarray.cmxa` after `... .opam/4.05.0+flambda/lib/hdf5_raw/hdf5_raw.cmxa`.  If I reverse their order, test.ml gets compiled successfully.  So I can compile test.ml if I do it by hand rather than letting jbuilder do it.

I haven't yet understood what controls the order of those library references in the compile line when jbuilder runs.  It's not my jbuild file, afaics.  

Since I can compile test.ml, which opens `Hdf5_caml`, I suspect now that the problem with `caml_copy_string` is just that utop can't make `caml_copy_string` available.  It makes sense that `caml_copy_string` is not the sort of thing that would normally be available in utop.  (Unless, maybe, one builds a custom utop that includes it?  If so, I might do that.  It will probably be useful to be able to experiment with `Hdf5_caml` in utop.)

**Comment by mars0i** [comment]
It doesn't seem to be a problem with utop.  My current solution is to use 0.1.3, following a suggestion here: https://discuss.ocaml.org/t/where-does-caml-copy-string-live/1260/4?u=mars0i

(It's great to be able to use HDF5.  Thanks.  Currently just experimenting for future serious use, trying to understand how it works.  I'd like to be able to use 0.1.4, though, since there are more functions implemented.)

**Comment by mars0i** [comment]
#11 fixes the problem with compiling to a native binary in the current version of the code.  If that PR seems OK, this issue could be closed, I think.

I still think it's valuable to be able to use `Hdf5_caml` in utop, which was possible in 0.1.3.  The current use of `caml_copy_string` in 0.1.4 prevents this, apparently.  I believe this is because `caml_copy_string` is unavailable in bytecode.  `caml_copy_string` is mentioned above and in #9, but those remarks are mixed in with other topics.  I'll submit an issue on that topic alone.  Hopefully some other way of accomplishing the same goal can be added in a future version of hdf5-ocaml so that it can be used in toplevels.

Thanks again for this library.

**Comment by vbrankov** [comment]
Merged the pull request #11 .

**Comment by mars0i** [comment]
Great--thanks.
